kamu-core/src/infra/utils/simple_transfer_protocol.rs

```diff
@@ -32,58 +37,141 @@ impl SimpleTransferProtocol {
         let src_chain = src.as_metadata_chain();
         let dst_chain = dst.as_metadata_chain();
 
-        let src_data = src.as_data_repo();
-        let dst_data = dst.as_data_repo();
+        let src_head = self.get_src_head(src_ref, src_chain).await?;
+        let dst_head = self.get_dest_head(dst_chain).await?;
+
+        info!(?src_head, ?dst_head, "Resolved heads");
 
-        let src_checkpoints = src.as_checkpoint_repo();
-        let dst_checkpoints = dst.as_checkpoint_repo();
+        let chains_comparison = MetadataChainComparator::compare_chains(
+            src_chain,
+            &src_head,
+            dst_chain,
+            dst_head.as_ref(),
+        )
+        .await?;
 
-        let src_head = match src_chain.get_ref(&BlockRef::Head).await {
+        match chains_comparison {
+            CompareChainsResult::Equal => return Ok(SyncResult::UpToDate),
+            CompareChainsResult::LhsAhead { .. } => { /* Skip */ }
+            CompareChainsResult::LhsBehind {
+                ref rhs_ahead_blocks,
+            } => {
+                if !force {
+                    return Err(SyncError::DestinationAhead(DestinationAheadError {
+                        src_head: src_head,
+                        dst_head: dst_head.unwrap(),
+                        dst_ahead_size: rhs_ahead_blocks.len(),
+                    }));
+                }
+            }
+            CompareChainsResult::Divergence {
+                uncommon_blocks_in_lhs: uncommon_blocks_in_src,
+                uncommon_blocks_in_rhs: uncommon_blocks_in_dst,
+            } => {
+                if !force {
+                    return Err(SyncError::DatasetsDiverged(DatasetsDivergedError {
+                        src_head: src_head,
+                        dst_head: dst_head.unwrap(),
+                        uncommon_blocks_in_dst,
+                        uncommon_blocks_in_src,
+                    }));
+                }
+            }
+        };
+
+        let blocks = match chains_comparison {
+            CompareChainsResult::Equal => unreachable!(),
+            CompareChainsResult::LhsAhead {
+                lhs_ahead_blocks: src_ahead_blocks,
+            } => src_ahead_blocks,
+            CompareChainsResult::LhsBehind { .. } | CompareChainsResult::Divergence { .. } => {
+                // Load all source blocks from head to tail
+                assert!(force);
+                self.map_block_iteration_errors(src_chain.iter_blocks().try_collect().await)?
```

Comment:
Might be more idiomatic to express this as:

```rust
src_chain
  .iter_blocks()
  .try_collect()
  .await
  .map_err(map_block_iteration_errors)?
````
This way `map_block_iteration_errors` won't even have to know about `Vec<T>`

Reply:
Nice!